examples/src/main/scala/org/apache/spark/examples/glrm/SparkGLRM.scala

```diff
@@ -0,0 +1,223 @@
+    val sr = regPen * stepSize
+    val arr = v.toArray.map(x =>
+      if (math.abs(x) < sr) 0
+      else if (x < -sr) x + sr
+      else x - sr
+    )
+    new BDV[Double](arr)
+  }
+
+  // Non-negative prox
+  def proxNonneg(v:BDV[Double], stepSize:Double, regPen:Double): BDV[Double] = {
+    val arr = v.toArray.map(x => math.max(x, 0))
+    new BDV[Double](arr)
+  }
+
+  /* End of GLRM libarry */
+
+
+  // Helper functions for updating
+  def computeLossGrads(ms: Broadcast[Array[BDV[Double]]], us: Broadcast[Array[BDV[Double]]],
+                       R: RDD[(Int, Int, Double)],
+                       lossGrad: (Int, Int, Double, Double) => Double) : RDD[(Int, Int, Double)] = {
+    R.map { case (i, j, rij) => (i, j, lossGrad(i, j, ms.value(i).dot(us.value(j)), rij))}
+  }
+
+  // Update factors
+  def update(us: Broadcast[Array[BDV[Double]]], ms: Broadcast[Array[BDV[Double]]],
+             lossGrads: RDD[(Int, Int, Double)], stepSize: Double,
+             nnz: Array[Double],
+             prox: (BDV[Double], Double, Double) => BDV[Double], regPen: Double)
+  : Array[BDV[Double]] = {
+    val rank = ms.value(0).length
+    val ret = Array.fill(ms.value.size)(BDV.zeros[Double](rank))
+
+    val retu = lossGrads.map { case (i, j, lossij) => (i, us.value(j) * lossij) } // vector/scalar multiply
+                .reduceByKey(_ + _).collect() // vector addition through breeze
+
+    for (entry <- retu) {
+      val idx = entry._1
+      val g = entry._2
+      val alpha = (stepSize / (nnz(idx) + 1))
+
+      ret(idx) = prox(ms.value(idx) - g * alpha, alpha, regPen)
+    }
+
+    ret
+  }
+
+  def fitGLRM(R: RDD[(Int, Int, Double)], M:Int, U:Int,
+              lossFunctionGrad: (Int, Int, Double, Double) => Double,
+              moviesProx: (BDV[Double], Double, Double) => BDV[Double],
+              usersProx: (BDV[Double], Double, Double) => BDV[Double],
+              rank: Int,
+              numIterations: Int,
+              regPen: Double) : (Array[BDV[Double]], Array[BDV[Double]], Array[Double]) = {
+    // Transpose data
+    val RT = R.map { case (i, j, rij) => (j, i, rij) }.cache()
+
+    val sc = R.context
+
+    // Compute number of nonzeros per row and column
+    val mCountRDD = R.map { case (i, j, rij) => (i, 1) }.reduceByKey(_ + _).collect()
+    val mCount = Array.ofDim[Double](M)
+    for (entry <- mCountRDD)
+      mCount(entry._1) = entry._2
+    val maxM = mCount.max
+    val uCountRDD = R.map { case (i, j, rij) => (j, 1) }.reduceByKey(_ + _).collect()
+    val uCount = Array.ofDim[Double](U)
+    for (entry <- uCountRDD)
+      uCount(entry._1) = entry._2
+    val maxU = uCount.max
+
+    // Initialize m and u
+    var ms = Array.fill(M)(BDV[Double](Array.tabulate(rank)(x => math.random / (M * U))))
+    var us = Array.fill(U)(BDV[Double](Array.tabulate(rank)(x => math.random / (M * U))))
+
+    // Iteratively update movies then users
+    var msb = sc.broadcast(ms)
+    var usb = sc.broadcast(us)
```

Comment:
Does it make assumption that both `ms` and `us` can fit on each of the executors? 
How well does it scale?
Thanks!